pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java

```diff
@@ -483,19 +486,42 @@ private ByteBuf applyCompression(ByteBuf payload) {
         return compressedPayload;
     }
 
+    /**
+     * Please ignore the note. below if you have a customized {@param callback}.
+     *
+     * Otherwise, please confirm the value "refCnt()" of {@param message} {@link MessageImpl#getDataBuffer}' is "2"
+     * when you call this method, because the ByteBuf will be released twice:
+     * 1. When enabled "isBatchMessagingEnabled":
+     *      1-1. Releasing when it was pushed into the batched message queue, see: {@link #doBatchSendAndAdd}
+     *      1-2. Releasing in the method {@link SendCallback#sendComplete(Throwable, OpSendMsgStats)}.
+     * 2. Single message, in other words, disabled "isBatchMessagingEnabled":
+     *      2-1. Releasing when it was writen out by
+     *           {@link ChannelOutboundHandler#write(ChannelHandlerContext, Object, ChannelPromise)}.
+     *      2-2. Releasing in the method {@link SendCallback#sendComplete(Throwable, OpSendMsgStats)}.
+     */
     public void sendAsync(Message<?> message, SendCallback callback) {
         checkArgument(message instanceof MessageImpl);
-
-        if (!isValidProducerState(callback, message.getSequenceId())) {
-            return;
-        }
-
         MessageImpl<?> msg = (MessageImpl<?>) message;
         MessageMetadata msgMetadata = msg.getMessageBuilder();
         ByteBuf payload = msg.getDataBuffer();
         final int uncompressedSize = payload.readableBytes();
 
+        // To guarantee compatibility with customized "SendCallback", see detail the doc of this method.
+        if ((callback instanceof ProducerImpl.DefaultSendMessageCallback
+            || callback.getClass().getName().endsWith("NonPersistentReplicator$ProducerSendCallback")
+            || callback.getClass().getName().endsWith("ProducerSendCallback$ProducerSendCallback"))
+            && msg.getDataBuffer().capacity() > 0) {
+            checkArgument(msg.getDataBuffer().refCnt() >= 2,
+                    "Message's data's refCnt is less than 2, see #23738");
+        }
```

Comment:
Such type assertions really should be avoided. I mean, the `checkArgument(message instanceof MessageImpl)` assertion before is also redundant.

The reference count validation should not be performed in `sendAsync`, instead, it should be validated before the caller calls `sendAsync` (e.g. in `GeoPersistentReplicator#replicateEntries`).

Reply:
Removed